File: backend/store-management/src/main/java/com/storemanagement/service/InventoryTransactionService.java
package com.storemanagement.service;

import com.storemanagement.dto.inventory.InventoryTransactionDTO;
import com.storemanagement.dto.PageResponse;
import com.storemanagement.utils.TransactionType;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;

public interface InventoryTransactionService {

    PageResponse<InventoryTransactionDTO> getAllTransactions(Pageable pageable);

    PageResponse<InventoryTransactionDTO> getTransactionsByProduct(Integer productId, Pageable pageable);

    PageResponse<InventoryTransactionDTO> getTransactionsByReference(
            com.storemanagement.utils.ReferenceType referenceType,
            Integer referenceId,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> getTransactionsByDateRange(
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> getTransactionsByProductAndDateRange(
            Integer productId,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> getTransactionsByType(
            TransactionType transactionType,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> getTransactionsByMultipleCriteria(
            TransactionType transactionType,
            Integer productId,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> getTransactionsByAdvancedCriteria(
            TransactionType transactionType,
            com.storemanagement.utils.ReferenceType referenceType,
            Integer productId,
            String productName,
            String sku,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable
    );

    PageResponse<InventoryTransactionDTO> filterTransactions(
            TransactionType transactionType,
            com.storemanagement.utils.ReferenceType referenceType,
            Integer referenceId,
            Integer productId,
            String productName,
            String sku,
            String brand,
            LocalDateTime fromDate,
            LocalDateTime toDate,
            Pageable pageable
    );
}

File: frontend_admin/src/services/inventoryTransactionService.js
import api from "./api";

const unwrap = (resp) => resp?.data?.data ?? resp?.data;

export const inventoryTransactionService = {

  getAllTransactions: async (params = {}) => {
    const {
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    const response = await api.get("/inventory-transactions", {
      params: queryParams,
    });
    return unwrap(response);
  },

  getTransactionsByProduct: async (productId, params = {}) => {
    const {
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    const response = await api.get(
      `/inventory-transactions/product/${productId}`,
      {
        params: queryParams,
      }
    );
    return unwrap(response);
  },

  filterTransactions: async (params = {}) => {
    const {
      transactionType,
      productId,
      startDate,
      endDate,
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    if (transactionType) queryParams.transactionType = transactionType;
    if (productId) queryParams.productId = productId;
    if (startDate) queryParams.startDate = startDate;
    if (endDate) queryParams.endDate = endDate;

    const response = await api.get("/inventory-transactions/filter", {
      params: queryParams,
    });
    return unwrap(response);
  },

  filterTransactionsAdvanced: async (params = {}) => {
    const {
      transactionType,
      referenceType,
      productId,
      productName,
      sku,
      startDate,
      endDate,
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    if (transactionType) queryParams.transactionType = transactionType;
    if (referenceType) queryParams.referenceType = referenceType;
    if (productId) queryParams.productId = productId;
    if (productName) queryParams.productName = productName;
    if (sku) queryParams.sku = sku;
    if (startDate) queryParams.startDate = startDate;
    if (endDate) queryParams.endDate = endDate;

    const response = await api.get("/inventory-transactions/filter-advanced", {
      params: queryParams,
    });
    return unwrap(response);
  },

  getTransactionsByReference: async (
    referenceType,
    referenceId,
    params = {}
  ) => {
    const {
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      referenceType,
      referenceId,
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    const response = await api.get("/inventory-transactions/reference", {
      params: queryParams,
    });
    return unwrap(response);
  },

  getTransactionsByType: async (transactionType, params = {}) => {
    const {
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      transactionType,
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    const response = await api.get("/inventory-transactions/by-type", {
      params: queryParams,
    });
    return unwrap(response);
  },

  getTransactionHistory: async (params = {}) => {
    const {
      startDate,
      endDate,
      productId,
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    if (startDate) queryParams.startDate = startDate;
    if (endDate) queryParams.endDate = endDate;
    if (productId) queryParams.productId = productId;

    const response = await api.get("/inventory-transactions/history", {
      params: queryParams,
    });
    return unwrap(response);
  },

  searchTransactions: async (params = {}) => {
    const {
      transactionType,
      referenceType,
      referenceId,
      productId,
      productName,
      sku,
      brand,
      fromDate,
      toDate,
      pageNo = 1,
      pageSize = 10,
      sortBy = "transactionDate",
      sortDirection = "DESC",
    } = params;

    const queryParams = {
      pageNo,
      pageSize,
      sortBy,
      sortDirection,
    };

    if (transactionType) queryParams.transactionType = transactionType;
    if (referenceType) queryParams.referenceType = referenceType;
    if (referenceId) queryParams.referenceId = referenceId;
    if (productId) queryParams.productId = productId;
    if (productName) queryParams.productName = productName;
    if (sku) queryParams.sku = sku;
    if (brand) queryParams.brand = brand;
    if (fromDate) queryParams.fromDate = fromDate;
    if (toDate) queryParams.toDate = toDate;

    const response = await api.get("/inventory-transactions/search", {
      params: queryParams,
    });
    return unwrap(response);
  },
};
